<a href="https://colab.research.google.com/github/praveenadanthapally/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

import pandas as pd
import numpy as np

# Load the starter dataset
DATA_PATH = "../../data/raw/content_refresh_anonymized.csv"

df = pd.read_csv(DATA_PATH)

print("Rows:", len(df))
print("Columns:", len(df.columns))

# Basic checks
print("\nFirst 5 rows:")
display(df.head())

# Week 1 discovery: pages with impressions
if "impressions_90d" in df.columns:
    positive_impressions = df[df["impressions_90d"] > 0]

    print("\nRows with impressions_90d > 0:", len(positive_impressions))

    if "search_volume" in df.columns:
        corr = positive_impressions[
            ["search_volume", "impressions_90d"]
        ].corr().iloc[0, 1]

        print(
            "Correlation between search_volume and impressions_90d:",
            round(corr, 3)
        )

# Starter label distribution
if "trend_direction" in df.columns:
    declining_rate = (
        df["trend_direction"].eq("down").mean()
    )

    print(
        "Declining rate:",
        round(declining_rate, 3)
    )

# Useful summary statistics for the chosen lane
summary_cols = [
    col for col in [
        "impressions_90d",
        "sessions_90d",
        "ctr",
        "avg_position",
        "content_age_days",
        "days_since_last_update",
        "word_count"
    ]
    if col in df.columns
]

print("\nSummary statistics:")
display(df[summary_cols].describe().T)

## 2. The question: decision, action, and cost of a wrong call

### Research question

**Which pages should be reviewed first for a potential content refresh or other content action, based on observable search and engagement signals?**

### Decision

The decision is which pages should receive limited review and optimization attention first.

### Action

A reviewer could inspect the highest-priority pages and decide whether to:

- refresh existing content;
- expand thin content;
- improve CTR-related elements;
- protect a page that is performing well;
- consider consolidation or pruning;
- or monitor the page instead of making a change.

### Cost of a wrong recommendation

A false positive could waste editorial or SEO effort on a page that did not need attention.

A false negative could cause a genuinely useful opportunity to be missed.

Because of these costs, the output should be a **ranked review queue**, not an automatic instruction to change a page.

In [7]:
# Clone the FlyRank internship repository
!git clone https://github.com/praveenadanthapally/flyrank-ml-internship.git

# Move into the repository
%cd flyrank-ml-internship

print("Repository ready.")

Cloning into 'flyrank-ml-internship'...
remote: Enumerating objects: 124, done.
remote: Counting objects: 100% (124/124), done.
remote: Compressing objects: 100% (79/79), done.
remote: Total 124 (delta 36), reused 97 (delta 28), pack-reused 0 (from 0)
Receiving objects: 100% (124/124), 1.86 MiB | 11.57 MiB/s, done.
Resolving deltas: 100% (36/36), done.
/content/flyrank-ml-internship
Repository ready.


In [8]:
import pandas as pd
import numpy as np

# Load the starter dataset from the repository
DATA_PATH = "data/raw/content_refresh_anonymized.csv"

df = pd.read_csv(DATA_PATH)

print("Rows:", len(df))
print("Columns:", len(df.columns))

print("\nFirst 5 rows:")
display(df.head())

# Week 1 discovery: pages with impressions
if "impressions_90d" in df.columns:
    positive_impressions = df[df["impressions_90d"] > 0]

    print("\nRows with impressions_90d > 0:", len(positive_impressions))

    if "search_volume" in df.columns:
        corr = positive_impressions[
            ["search_volume", "impressions_90d"]
        ].corr().iloc[0, 1]

        print(
            "Correlation between search_volume and impressions_90d:",
            round(corr, 3)
        )

# Starter label distribution
if "trend_direction" in df.columns:
    declining_rate = df["trend_direction"].eq("down").mean()

    print(
        "Declining rate:",
        round(declining_rate, 3)
    )

# Useful summary statistics for the chosen lane
summary_cols = [
    col for col in [
        "impressions_90d",
        "sessions_90d",
        "ctr",
        "avg_position",
        "content_age_days",
        "days_since_last_update",
        "word_count"
    ]
    if col in df.columns
]

print("\nSummary statistics:")
display(df[summary_cols].describe().T)

Rows: 30000
Columns: 44

First 5 rows:


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7



Rows with impressions_90d > 0: 30000
Correlation between search_volume and impressions_90d: 0.001
Declining rate: 0.542

Summary statistics:


,count,mean,std,min,25%,50%,75%,max
impressions_90d,30000.0,5200.366300,16838.019547,1.0,81.0,731.00,3615.25,517715.0
sessions_90d,30000.0,37.066633,107.069131,1.0,2.0,7.00,27.00,4345.0
ctr,30000.0,0.510733,3.279162,0.0,0.0,0.07,0.29,100.0
avg_position,30000.0,16.342380,15.216790,0.0,6.2,10.80,22.30,245.0
content_age_days,30000.0,256.167800,132.707930,90.0,132.0,236.00,333.00,564.0
days_since_last_update,30000.0,46.098300,42.078709,1.0,20.0,20.00,104.00,373.0
word_count,22301.0,3107.760325,1452.382598,8.0,2413.0,2877.00,3666.00,9546.0


## 3. Quick look at the data

The code cell above loads the starter dataset and checks several observable signals relevant to the refresh opportunity question.

The main numbers I will use as evidence are:

- **Rows in the starter dataset:** 30000
- **Rows with impressions_90d > 0:** 0.01
- **Declining rate:** 0.542

The dataset also contains signals such as impressions, sessions, CTR, average position, content age, days since last update, and word count.

These numbers make the lane worth investigating because there is measurable search visibility and content-performance information that can potentially be used to prioritize a limited review queue.

The numbers are descriptive evidence from the starter dataset. They do not by themselves show that changing a page will cause better performance.

## 4. Careful words: what I can and can't claim

### What I can claim

I can describe associations and patterns observed in the available data.

I can test whether a simple ranking or model is useful for prioritizing pages for human review.

I can compare a transparent baseline with a learned method using an appropriate validation design and ranking metrics such as Precision@K.

### What I cannot claim

I cannot claim that a particular signal is a proven Google ranking factor.

I cannot claim that a page identified as high priority is guaranteed to improve after a refresh.

I cannot claim that a refresh caused a recovery using this observational dataset alone.

I also should not treat a model score as the truth. The score is a prioritization aid that should support human review.

For this project, I will use careful language such as **"associated with," "observed," "suggests," and "useful for prioritization"** rather than making causal claims.

## 5. Self-check

- [x] I selected one of the predefined lanes.
- [x] I stated the research question.
- [x] I identified the decision being improved.
- [x] I described the action a reviewer could take.
- [x] I explained the cost of false positives and false negatives.
- [x] I loaded the starter dataset in a code cell.
- [x] I included real numbers from the starter dataset.
- [x] I explained why data or ML could help with prioritization.
- [x] I treated the problem as decision support rather than guaranteed automation.
- [x] I avoided claiming that a refresh causes better performance.
- [x] I avoided claiming that the data proves Google ranking factors.